File: BoxMuller.py
#-------------------------------
# Box-Muller transform in Python
#-------------------------------

import numpy as np
import pandas as pd
import math 

# 1. generate uniform vectors
n = 10000
np.random.seed(2023)
u1 = np.random.uniform(low=0.0, high=1.0, size=n)
u2 = np.random.uniform(low=0.0, high=1.0, size=n)

# 2. define Box-Müller transform using vectorization
x1 = (np.sqrt(-2*np.log(u1))*np.cos(2*math.pi*u2))
x2 = (np.sqrt(-2*np.log(u1))*np.sin(2*math.pi*u2))
x = pd.DataFrame({'x1':x1, 'x2':x2})

# 3. print first lines of dataset 
data = x
data[0:6]
#          x1        x2
# 0 -0.860024  1.235653
# 1 -0.341486 -0.339862
# 2 -0.577473  0.853466
# 3  1.963577 -0.527138
# 4 -1.889265  0.586374
# 5  1.231823 -0.040409

# 4. plot the standard normal data

import seaborn as sns
sns.jointplot(data=data, x="x1", y="x2")


#----
# end
#----

File: BoxMuller.R
#-----------------------
# Box-Muller transform
#-----------------------

# 1. generate uniform vectors
n = 10000
set.seed(2023)
u1 = runif(n)
u2 = runif(n)

# 2. define Box-Müller transform function
x1 = sqrt(-2*log(u1))*cos(2*pi*u2)
x2 = sqrt(-2*log(u1))*sin(2*pi*u2)
x = data.frame(x1, x2)

# 3. print first lines of dataset 
data = x
data[1:6,]
#          x1         x2
# 1  1.139567 -0.4752820
# 2 -1.449969  0.2893102
# 3 -1.791041  0.6499395
# 4 -1.255448  0.5252876
# 5  2.625178 -0.3092095
# 6  1.735641 -1.1015367

# 4. plotting

library(ggplot2)

p1<- ggplot(data, aes(x = x1)) + 
  geom_histogram(aes(y = ..density..),
                 colour = 1, fill = "white") +
  geom_density(lwd = 1, colour = "#a50101") + theme_minimal()

p2<- ggplot(data, aes(x = x2)) + 
  geom_histogram(aes(y = ..density..),
                 colour = 1, fill = "white") +
  geom_density(lwd = 1, colour = "#a50101") + theme_minimal()

p3<- ggplot(data, mapping = aes(x = x1, y = x2)) +
  geom_point(size = 0.6) + 
  geom_jitter(alpha = 0.75, size = 0.6) + theme_minimal()

library(gridExtra)

grid.arrange(arrangeGrob(p1, p2), 
             arrangeGrob(p3, ncol=1), 
             ncol=2, widths=c(1,1))

#----
# end
#----
